bddeb: new --packaging-branch argument to pull packaging from branch

paride opened the pull request:
bddeb builds a .deb package using the template packaging files in
packages/debian/.

The new --packaging-branch flag allows to specify a git branch
where to pull the packaging (i.e. the debian/ directory) from.
This is useful to build a .deb package from master with the very
same packaging which is used for the uploads.

paride commented:
I was temped to propose this as the new default build method, but the "templated" approach still has a couple of advantages:

- doesn't require a packaging branch
- auto generates the package dependencies via `read-dependencies`

and possibly other things I'm missing, so I'm leaving the default alone.

raharper commented on packages/bddeb:
```diff
@@ -99,6 +100,27 @@ def write_debian_folder(root, templ_data, cloud_util_deps):
+    """Import a debian package directory from a branch."""
+    print("Importing debian/ from branch %s to %r" % (branch, root))
+    p_dumpdeb = subprocess.Popen(
+        ["git", "archive", branch, "debian"], stdout=subprocess.PIPE
```
if your target branch does not have a directory 'debian' this will fail, do you want to either precheck or handling the expected failure ?

raharper reviewed (commented):
I really like this change.  For older release branches (like xenial) it's much harder to test correctly since bddeb won't have many of the per-branch changes to the runtime code (not just packing changes).

Some comments in-line

paride commented:
Thanks for the comments @raharper, I updated the branch. Notable changes:
 - more use of `subp()`
 - check if `debian/control` exists in the specified branch, per your suggestion
 - renamed `--packaging-from-branch` to `--packaging-branch`
 - exclude `debian/patches` when importing the packaging

I'd like to have some feedback on the `debian/patches` exclusion. Seems a good idea as bddeb is mainly used from master, and as master moves patches do not apply anymore, and the build breaks. And by default (i.e. without `--packaging-branch`) bddeb doesn't even know about patches. However when building from *release* branches it would be nice to include the patches. I thought about not excluding the patches if `pakaging_branch == current_branch`, but this makes the script behavior less obvious. Thoughts?

raharper commented:
> I'd like to have some feedback on the `debian/patches` exclusion. Seems a good idea as bddeb is mainly used from master, and as master moves patches do not apply anymore, and the build breaks. And by default (i.e. without `--packaging-branch`) bddeb doesn't even know about patches. However when building from _release_ branches it would be nice to include the patches. I thought about not excluding the patches if `pakaging_branch == current_branch`, but this makes the script behavior less obvious. Thoughts?

I very much would expect patches to come as well; otherwise I'd just build master.  You're correct that building master with merged release branch and patches won't always work.  However, our daily recipes already have to deal with this;  IIUC, there is a two step branch merge to address this.  @blackboxsw and @OddBloke have this worked out for our daily recipe; so we'd need to pull that in.

Given that it's more complicated, but desirable (IMO) then I think this could use a --without-patches to disable this behavior.

paride commented:
If I connected the dots correctly correctly, the daily recipe reverts the cherry-picked patches, which are identified by their filename (has to match `debian/patches/*cpick*`). We could exclude the cherry-picks instead, instead of excluding all the patches.

paride commented:
That's probably not a good idea either. I dropped the "exclude patches" bit, let's wait and see if we hit an actual situation where we would like some patches to be excluded and then see if there's a sensible way to do it.

paride commented:
Rebased on master.

paride commented:
hey @raharper, does the current PR address your review suggestions?

raharper commented on cloudinit/subp.py:
```diff
@@ -181,6 +181,8 @@ def subp(args, data=None, rcs=None, env=None, capture=True,
         call this fuction with a single string argument before starting
         and after finishing.
+    :param cwd:
+        changee the working directory to cwd before executing the command.
```
```suggestion
        change the working directory to cwd before executing the command.
```

raharper commented on packages/bddeb:
```diff
@@ -21,6 +22,7 @@ def find_root():
     raise OSError(("Unable to determine where your cloud-init topdir is."
                    " set CLOUD_INIT_TOP_D?"))
 
+
```
Drop

paride replied:
That's actually a pylint drive-by fix, it requires 2 blank lines there. Are you for dropping as it's unrelated??

paride replied:
Well, dropped. I'll fix linting in a separate PR.

raharper commented on packages/bddeb:
```diff
@@ -99,6 +101,36 @@ def write_debian_folder(root, templ_data, cloud_util_deps):
 
+def write_debian_folder_from_branch(root, templ_data, branch):
+    """Import a debian package directory from a branch."""
+    print("Importing debian/ from branch %s to %r" % (branch, root))
```
```suggestion
    print("Importing debian/ from branch %s to %s" % (branch, root))
```

paride replied:
The original `write_debian_folder()` I took this from uses `%r`, but `root` is a string, so +1 for the change.

raharper commented on packages/bddeb:
```diff
@@ -99,6 +101,36 @@ def write_debian_folder(root, templ_data, cloud_util_deps):
+
+    print("Adding new entry to debian/changelog")
+    full_deb_version = (
+        templ_data["version_long"] + "-1~bddeb" + templ_data["release_suffix"]
```
Did we close on whether this value ("-1~bddeb") needs to be user-controllable?

paride replied:
Like this is behaves like plain `bddeb` without `--packaging-branch`. We can certainly make it configurable, but is it useful in practice?

blackboxsw commented:
> That's probably not a good idea either.
> 
> I dropped the "exclude patches" bit, let's wait and see if we hit an actual situation where we would like some patches to be excluded and then see if there's a sensible way to do it.

I think that's the best option @paride.   We will have this case any time we have a "cpicked" patch back into an ubunut/<release>. We'll need to pop that cpick file off the stack if we want to continue to build using ubuntu/xenial/debian on master. I think this case is unneeded though as this complexity is solved by the ubuntu/daily/<release> branches and doesn't really need to be solved by your switch here. I think the majority of cases we would use this in would be just building master with ubuntu/devel/debian dir, and the devel branch shouldn't really carry cherry picks as we would mostly prefer to just release tip of master into ubuntu/devel for two reasons:
 * master is almost always releasable
 * we only cherry-pick into ubuntu/devel during devel series feature-freeze events once every 6 months.

I don't think it warrants us designing the cpick* patch-popping algorithm in bddeb as we don't use bddeb in the daily recipes. We instead rely on a simplified package branching strategy to allow our daily builds to succeed when merging tip of master into the ubuntu/<release> branch.

blackboxsw commented on packages/bddeb:
```diff
@@ -140,13 +171,52 @@ def get_parser():
+            subp.subp(
+                [
+                    "git",
+                    "show-ref",
```
This will only match/succeed if you actually have a local copy of the local branch checked out. Shouldn't we be checking upstream/args.packaging_branch ?

paride replied:
(You mean "a local copy of the *packaging* branch checked out" I guess.)

You are right, and the `git archive <branch> debian` command will also fail if `<branch>` (usually `ubuntu/devel`) is not checked out locally. We could try to do it automatically, but there are two caveats:
 * We don't know what's the remote the user wants to use. We could add heuristics at cost of extra complexity, probably not worth it.
 * I wouldn't like a "build package" tool to touch my git repo.

I'm thinking of just hinting the user to checkout the packaging branch in the error message. It's an easy step to do and also makes very clear where the packaging bits are coming from. What do you think?

blackboxsw replied:
I think the hinting sounds good.

paride commented:
Thanks for the review, @blackboxsw hint added.

I agree `ubuntu/devel` is the the branch that `--packaging-branch` is meant to target most of the time, and it's indeed the default, and we shouldn't normally carry cherry-picks there for the reasons you stated.
